url1a41b8e9f2_mr1azl_INF345_tgz-pJ8-TestParserJ8/src/xx/XPathEvaluateur.java

```diff
@@ -0,0 +1,362 @@
+package xx;
+
+
+import java.io.File;
+import java.io.IOException;
+import java.io.StringReader;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerConfigurationException;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+
+import org.w3c.dom.Attr;
+import org.w3c.dom.Document;
+import org.w3c.dom.DocumentType;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+import yyy.NodeTest;
+import yyy.Step;
+import zz.TestPT;
+
+ 
+public class XPathEvaluateur {
+
+	public XPathEvaluateur() {
+	}
+private String docFile="";
+
+/**
+ * 
+ * @param xpath
+ * @param document
+ * @return
+ */
+public Node compile(String xpath,Document document) {
+	try {
+		List<Node> dedup=new ArrayList<>();
+		DocumentBuilderFactory factory = DocumentBuilderFactory
+				.newInstance();
+		factory.setNamespaceAware(true);
+		factory.setValidating(false);
+		DocumentBuilder dBuilder = DocumentBuilderFactory.newInstance()
+				.newDocumentBuilder();
+
+		Element root = document.createElement("answer");
+//		root.setAttribute("expression", xpath);
+		Object result = Parser.parseExpression(xpath, new MonCompiler(
+				document));
+		if (result instanceof String)
+			System.out.println("WRONG PARSING " + (String) result);
+		
+		 if (result instanceof NodeTest)
+			{//TODO
+				
+			}else
+			if (result instanceof Node){
+				 
+			//root.appendChild((Node) result);
+			Node node = (Node) result;
+			if(!dedup.contains(node) && ! (node instanceof DocumentType)){
+				root.appendChild(node.cloneNode(true));
+				dedup.add(node);
+				}
+		}
+		if (result instanceof NodeList)
+			for (int i = 0; i < ((NodeList) result).getLength(); i++)
+				root.appendChild(((NodeList) result).item(i));
+		if (result instanceof ArrayList)
+			for (Node node : (ArrayList<Node>) result) {
+				if (node instanceof Document) {
+					System.out.println("the node is document");
+					printNode(node, "\t");
+//					root = ((Document) node).getDocumentElement();
+				} else if (node instanceof Attr) {
+					root.setAttribute(((Attr) node)
+							.getName(),((Attr) node)
+							.getValue());
+//					root.appendChild(document.createElement(((Attr) node)
+//							.getName()));
+				}else if (node instanceof Step)
+				{
+					
+				}
+				else {
+					if(!dedup.contains(node) && ! (node instanceof DocumentType)){
+						root.appendChild(node.cloneNode(true));
+						dedup.add(node);
+						}
+				}
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 82:104, 152:174.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.